### PR TITLE
AIML-481: Security hardening — HTTPS enforcement, error sanitization, header redaction

### PR DESCRIPTION
**Pull request by ChrisEdwards**
## Summary
- Reject HTTP at startup so plaintext credential transmission is impossible
- Replace raw `e.getMessage()` in client-facing errors with opaque text + log-correlation `requestId`
- Strip `Authorization`, `Cookie`, and `Set-Cookie` from `get_vulnerability` HTTP request output

## Why This Change Exists

A Feb 2026 security review of the Contrast MCP server (Jira [AIML-481](https://contrast.atlassian.net/browse/AIML-481)) evaluated 28 security-tagged items. This PR addresses three validated **Credential / Secret Protection** findings — places where the server could leak sensitive data to MCP clients (and by extension, to whatever LLM is consuming them):

1. **Insecure protocol allowed.** `SDKHelper.getProtocolAndServer()` accepted any string — `http`, `ftp`, anything — and prepended it to the hostname. A single misconfiguration of `contrast.api.protocol=http` would have transmitted API key, service key, and session cookies in plaintext. There was no validation at startup.
2. **Raw exception messages forwarded to clients.** `SingleTool` / `PaginatedTool` catch blocks returned `"Internal error: " + e.getMessage()` to the agent. Those messages can carry internal API paths (`/api/ng/org-abc123/traces`), org IDs, and stack-trace fragments that should never reach an LLM.
3. **Live credentials in HTTP request output.** `get_vulnerability` returned the full captured HTTP request including `Authorization: Bearer …` and `Cookie: …` headers — live tokens from the monitored application, not from Contrast itself.

## Approach We Chose

Three targeted fixes rather than a broad lockdown:

- **Protocol enforcement at two layers.** Startup-time validation in `ContrastSDKFactory.validateConfiguration()` plus defense-in-depth in `SDKHelper.getProtocolAndServer()`. Both case-insensitive.
- **Opaque errors with log correlation.** Replace `e.getMessage()` with `"An internal error occurred (ref: <requestId>)"` for the agent; full exception (with stack) goes to the server log under the same `requestId` so operators can still debug.
- **Surgical header redaction.** `HttpRequestRedactor` removes only the three dangerous headers (and any RFC 7230 continuation lines) — URL, method, other headers, and body are preserved because they're useful for understanding the vulnerability.

Tradeoff: the `requestId` correlation pattern means agents see a less helpful error message. We chose this because the alternative — a curated allow-list of "safe" exception text — is fragile and easy to regress.

## Outcome of This Change

- Server **refuses to start** if `contrast.api.protocol` is anything other than `https` (or blank/null), or if `CONTRAST_HOST_NAME` carries an `http://` scheme. Clear error messages name the offending value.
- **Zero** raw exception messages reach MCP clients. All error and warning paths sanitized; tests assert sensitive substrings (API paths, org IDs) are absent.
- `get_vulnerability` redacts sensitive auth headers automatically. `httpRequest` field description updated to disclose the redaction.
- 670 unit tests pass; checkstyle + format clean.

No production metrics gathered — this is preventive hardening, not a fix for an observed leak.

## Reviewer Guide

1. **Start at the redactor:** `HttpRequestRedactor.java` (new) and how `GetVulnerabilityTool.java` calls it.
2. **Then protocol validation:** `ContrastSDKFactory.validateConfiguration()` and the parallel checks in `SDKHelper.getProtocolAndServer()`.
3. **Then error sanitization:** `PaginatedTool.java` / `SingleTool.java` catch blocks, plus the warning paths in `GetVulnerabilityTool.java`, `ListApplicationsByCveTool.java`, and `MetadataJsonFilterSpec.java`.
4. **Tests last:** `HttpRequestRedactorTest`, `ContrastSDKFactoryTest`, and the negative assertions added to `PaginatedToolTest` / `SingleToolTest`.

### 1. Problem framing
Three independent leak vectors, all in the path between Contrast SDK ↔ MCP client. Fix where each surfaces.

### 2. Core implementation

**HTTPS-only enforcement** lives in two places:

- `ContrastSDKFactory.validateConfiguration()` (`@PostConstruct`) — strips + lowercases the configured protocol; rejects `https://` (common typo where the user pastes a URL into the property), rejects any non-`https` value with an `IllegalStateException` that names the offending value. Also rejects an `http://` scheme embedded in `CONTRAST_HOST_NAME`.
- `SDKHelper.getProtocolAndServer()` — same logic at URL-construction time, so even direct callers can't bypass validation. Throws `IllegalArgumentException`.

**Error sanitization** is uniform across tool base classes:

```java
// SingleTool.java / PaginatedTool.java
} catch (Exception e) {
  log.atError().addKeyValue("requestId", requestId).setCause(e).setMessage("Request failed unexpectedly").log();
  return SingleToolResponse.error("An internal error occurred (ref: " + requestId + ")");
}
```

`ResourceNotFoundException` simplified from `"Resource not found: " + e.getMessage()` to plain `"Resource not found"` — the SDK's message can contain the full API path.

In `GetVulnerabilityTool.java`, three optional fetches (recommendation, HTTP request, stack trace) had warnings of the form `"X data not available: " + e.getMessage()`. These now log the cause structurally and emit a fixed warning string to the client.

**Header redaction** in `HttpRequestRedactor.java`:

- Splits on `\r?\n`, preserves whichever line ending the input used.
- Treats lines until the first blank line as headers; everything after is the body and passes through untouched.
- Drops lines whose RFC 7230 token-chars-only header name (case-insensitively) matches `authorization` / `cookie` / `set-cookie`.
- Tracks `skipContinuation` so that any whitespace-prefixed RFC 7230 continuation line that follows a redacted header is also dropped.

### 3. Supporting changes
- `application.properties` comment updated: HTTP no longer supported.
- `GetVulnerabilityTool` `httpRequest` field description discloses redaction (so the agent doesn't hallucinate that headers were missing in the source data).
- `PR_TEMPLATE.md` and `scripts/br-set-design` are unrelated repo-tooling additions that landed in the same commit chain (`docs:` commits). Not load-bearing for the security work.

### 4. Edge cases and safeguards
- Protocol validation: `https://` typo, `HTTP`/`Http`/`http` casing variations, `ftp://`, blank, null, embedded scheme in hostname — all covered explicitly.
- Redactor: empty/null input returns input as-is; request with no headers (just a request line + body) is preserved verbatim; mixed `\r\n` and `\n` not assumed; case-insensitive match.
- Error sanitization: tests verify that known-leaky strings (org IDs, API paths) appear in the **logger** but not in the response.

### 5. How the tests prove it
- `ContrastSDKFactoryTest` — 12 cases drive `validateConfiguration()` through every accept/reject path; assertions check the exception type and that the message names the offending value.
- `HttpRequestRedactorTest` — golden inputs covering all sensitive headers, mixed casing, continuation lines, body preservation, no-header input, blank input.
- `PaginatedToolTest` / `SingleToolTest` — added assertions that the response error string does **not** contain SDK-internal substrings, and that it **does** contain the correlation `requestId`.
- `SDKHelperTest` — flipped previously-passing HTTP cases to expect `IllegalArgumentException`; added new cases for `https://` typo and case-insensitive matching.

### Production code
- `src/main/java/.../config/ContrastSDKFactory.java` — `@PostConstruct` protocol + hostname startup validation.
- `src/main/java/.../sdkextension/SDKHelper.java` — Reject HTTP / non-HTTPS at URL-construction time; case-insensitive.
- `src/main/java/.../tool/base/PaginatedTool.java` — Opaque error message with `requestId`; drop `e.getMessage()` from `ResourceNotFoundException` text.
- `src/main/java/.../tool/base/SingleTool.java` — Same as above.
- `src/main/java/.../tool/vulnerability/GetVulnerabilityTool.java` — Apply `HttpRequestRedactor` to captured request text; strip `e.getMessage()` from optional-fetch warnings; updated `httpRequest` field description.
- `src/main/java/.../tool/vulnerability/HttpRequestRedactor.java` — **New.** Stateless header redactor.
- `src/main/java/.../tool/library/ListApplicationsByCveTool.java` — Strip `e.getMessage()` from library-fetch warning.
- `src/main/java/.../tool/validation/MetadataJsonFilterSpec.java` — Strip parse-error detail from validation errors; log structured cause.
- `src/main/resources/application.properties` — Comment now says HTTPS-only.

### Tests
- `src/test/java/.../config/ContrastSDKFactoryTest.java` — **+118 lines.** Protocol/hostname startup-validation coverage.
- `src/test/java/.../tool/vulnerability/HttpRequestRedactorTest.java` — **New.** Full redactor coverage.
- `src/test/java/.../tool/base/PaginatedToolTest.java` / `SingleToolTest.java` — Negative assertions for sensitive substrings; positive assertions for `requestId` correlation.
- `src/test/java/.../sdkextension/SDKHelperTest.java` — Updated existing cases to match HTTPS-only policy; added new edge cases.
- 9 other test files — Mechanical updates so existing error-message assertions match the new sanitized format (`"Internal error: ..."` → `"An internal error occurred (ref: ..."`).

### Repo tooling (unrelated to security findings)
- `PR_TEMPLATE.md` — New PR template referenced by `CLAUDE.md`.
- `scripts/br-set-design` — Helper for `br update --design` with multi-line content.
- `CLAUDE.md` — Adds a "Helper Scripts" section pointing at `br-set-design`.
- `.gitignore` — Add `docs/brainstorms/`.

## Testing
- **Automated:** 670 unit tests pass; `make check-test` clean (checkstyle, Spotless, JUnit). Integration tests not re-run on this branch — they need credentials in `.env.integration-test`. CI on the parent PR exercises the integration suite.
- **Manual:** Verified server fails fast with `contrast.api.protocol=http` (clear startup error). Verified `HttpRequestRedactor` against captured request samples exercising `Authorization`, `Cookie`, `Set-Cookie`, and a continuation-line case.
- **Gaps:** No end-to-end test against a live Contrast org confirming the redacted HTTP request renders correctly to an MCP client. Header allow-list isn't externalized — additions require code change + deploy.

## Risks / Rollout / Follow-ups
- **Risk:** Operators with existing `contrast.api.protocol=http` configs will see startup failures after upgrade. Mitigated by clear error messages naming the offending value. No silent behavior change.
- **Risk:** The opaque error messages reduce agent debuggability. Mitigation: `requestId` correlates to server logs, which retain full context.
- **Follow-ups (out of scope, tracked under AIML-481 epic):** Supply-chain integrity (pin Maven wrapper / Docker base / GitHub Actions / wizcli), data minimization (cap unbounded responses), log hygiene (avoid INFO-level username/host).

---

Beads: `mcp-2tsm` (epic), `mcp-24s9` (HTTPS), `mcp-21rd` (sanitization), `mcp-1v4u` (redaction)

[AIML-481]: https://contrast.atlassian.net/browse/AIML-481?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/com/contrast/labs/ai/mcp/contrast/tool/vulnerability/HttpRequestRedactor.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.ArrayList;
+import java.util.Set;
+import org.springframework.util.StringUtils;
+
+final class HttpRequestRedactor {
+
+  private static final Set<String> SENSITIVE_HEADERS =
+      Set.of("authorization", "cookie", "set-cookie");
+
+  private HttpRequestRedactor() {}
```
Comment (dougj-contrast) [inline src/main/java/com/contrast/labs/ai/mcp/contrast/tool/vulnerability/HttpRequestRedactor.java:27]: you can use lombok for this boilerplate private constructor:
```
@NoArgsConstructor(access = PRIVATE)
```

File: src/main/java/com/contrast/labs/ai/mcp/contrast/config/ContrastSDKFactory.java [hunk 2]
```diff
@@ -98,5 +100,37 @@ public void validateConfiguration() {
           "Missing required Contrast configuration. Set the following environment variables: "
               + String.join(", ", missing));
     }
+
+    var protocol = properties.protocol();
```
Comment (dougj-contrast) [inline src/main/java/com/contrast/labs/ai/mcp/contrast/config/ContrastSDKFactory.java:104]: posted this in the other stacked PR
https://github.com/Contrast-Security-OSS/mcp-contrast/pull/85#discussion_r3149144400

File: src/main/java/com/contrast/labs/ai/mcp/contrast/tool/vulnerability/HttpRequestRedactor.java (hunk 1 above)
Comment (ChrisEdwards) [reply src/main/java/com/contrast/labs/ai/mcp/contrast/tool/vulnerability/HttpRequestRedactor.java:27]: Done — applied in #100 (stacked on #85). Class now uses `@NoArgsConstructor(access = AccessLevel.PRIVATE)` instead of the explicit private constructor.

File: src/main/java/com/contrast/labs/ai/mcp/contrast/config/ContrastSDKFactory.java (hunk 2 above)
Comment (ChrisEdwards) [reply src/main/java/com/contrast/labs/ai/mcp/contrast/config/ContrastSDKFactory.java:104]: Done — applied in #100 (stacked on #85). `validateConfiguration()` now builds a candidate URL from `protocol + hostName` and validates it via `new UrlValidator(new String[]{"https"}, ALLOW_LOCAL_URLS)`. Added `commons-validator:1.10.1` (past the 7-day soak window per ENTSEC-1742). Behavior preserved for every input the existing tests cover; see the PR walkthrough table for parity details.

File: src/main/java/com/contrast/labs/ai/mcp/contrast/tool/library/ListApplicationsByCveTool.java [hunk 3]
```diff
@@ -142,12 +142,12 @@ private void enrichAppsWithClassUsage(
...
         }
       } catch (Exception e) {
-        log.debug("Could not fetch library data for app {}: {}", app.getAppId(), e.getMessage());
-        warnings.add(
-            "Could not fetch class usage data for application '"
-                + app.getName()
-                + "': "
-                + e.getMessage());
+        log.atWarn()
+            .addKeyValue("appId", app.getAppId())
```
Comment (dougj-contrast) [inline src/main/java/com/contrast/labs/ai/mcp/contrast/tool/library/ListApplicationsByCveTool.java:146]: in a future PR you could add a `LoggingKeys` class with logging key constants so they can be re-used. I would imagine `appId` is something that will be used over and over

File: src/main/java/com/contrast/labs/ai/mcp/contrast/tool/vulnerability/GetVulnerabilityTool.java [hunk 4]
```diff
@@ -128,27 +128,39 @@ private VulnerabilityContext buildVulnerabilityContext(
...
       var requestResponse = sdk.getHttpRequest(orgId, vulnId);
       if (requestResponse != null && requestResponse.getHttpRequest() != null) {
-        httpRequestText = requestResponse.getHttpRequest().getText();
+        httpRequestText = HttpRequestRedactor.redact(requestResponse.getHttpRequest().getText());
       }
     } catch (Exception e) {
-      log.debug("Could not fetch HTTP request for {}: {}", vulnId, e.getMessage());
-      warnings.add("HTTP request data not available: " + e.getMessage());
+      log.atWarn()
+          .addKeyValue("vulnId", vulnId)
```
Comment (dougj-contrast) [inline src/main/java/com/contrast/labs/ai/mcp/contrast/tool/vulnerability/GetVulnerabilityTool.java:147]: another example logging `vulnId` used in multiple log statements...

example:
https://github.com/Contrast-Security-Inc/aiml-services/blob/develop/services/aiml-remediation-api/src/main/java/com/contrastsecurity/aiml/remediation/util/LoggingKeys.java

File: src/main/resources/application.properties [hunk 5]
```diff
@@ -15,6 +15,6 @@ contrast.user-name=${CONTRAST_USERNAME:}
 contrast.org-id=${CONTRAST_ORG_ID:}
 contrast.http-proxy-host=${http.proxy.host:${http_proxy_host:}}
 contrast.http-proxy-port=${http.proxy.port:${http_proxy_port:}}
-# This can be http or https. It should only be http if you are running teamserver locally.
+# Only HTTPS is supported. The server will refuse to start with HTTP or other protocols.
```
Comment (dougj-contrast) [inline src/main/resources/application.properties:18]: So how would you run this app locally to test without a cert?

**Review by dougj-contrast — APPROVED** [review]

File: src/main/resources/application.properties (hunk 5 above)
Comment (ChrisEdwards) [reply src/main/resources/application.properties:18]: This is the TeamServer URL. I don't know when you would want to test it against a local TS. That is a lot of work to spin up, and then there is no data there, so you would have to seed it. If that was something devs wanted to do, we could relax this. I won't ever run it local...too much work for no benefit.

File: src/main/java/com/contrast/labs/ai/mcp/contrast/tool/library/ListApplicationsByCveTool.java (hunk 3 above)
Comment (ChrisEdwards) [reply src/main/java/com/contrast/labs/ai/mcp/contrast/tool/library/ListApplicationsByCveTool.java:146]: Done — implemented in #100 (commit d8f9442). Added `LoggingKeys` class in `tool/base/` with constants for the keys we actually use (`requestId`, `durationMs`, `httpStatus`, `exceptionType`, `errorCount`, `itemCount`, `totalItems`, `filterName`) plus `appId` and `vulnId` per your suggestion. Refactored `PaginatedTool`, `SingleTool`, and `MetadataJsonFilterSpec` to use the constants.

File: src/main/java/com/contrast/labs/ai/mcp/contrast/tool/vulnerability/GetVulnerabilityTool.java (hunk 4 above)
Comment (ChrisEdwards) [reply src/main/java/com/contrast/labs/ai/mcp/contrast/tool/vulnerability/GetVulnerabilityTool.java:147]: Done — implemented in #100 (commit d8f9442). `LoggingKeys` class added with `VULN_ID` (and `APP_ID`) per your suggestion, plus the keys we currently use repeatedly. Style follows the aiml-services reference you linked.
